### PR TITLE
Allow direct imports to reduce bundle size.

### PR DESCRIPTION
**Pull request by ruiaraujo**
Webpack 2 does not remove a lot of unused modules from my bundle so direct import remain the best solution to reduce bundle size.

The doc page was heavily inspired by https://github.com/ReactTraining/react-router/blob/v3/docs/guides/MinimizingBundleSize.md

### CODE REVIEW

**Comment by gustavohenke** [comment]
Can you check #2315, please?
Over there guys think the bundle size problem is related to lodash.

BTW, what a huge PR!

**Comment by ruiaraujo** [comment]
For me since I was also using lodash, it is shared so that is not a problem. The problem was the factory index files that prevented direct imports.

**Comment by erikras** [comment]
Wow! That's an impressive PR!

Well done. Thank you for your time.

File: docs/faq/BundleSize.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+```js
+import reduxForm from 'react-router/es/reduxForm'
+import Field from 'react-router/es/Field'
+import FieldArray from 'react-router/es/FieldArray'
+```
+
+
+Use `es` if you are using a bundler that can process ES modules like `webpack@2` or `rollup`, otherwise use `lib`.
+
+The public API available in this manner is defined as the set of imports available from the top-level `react-router` module.
```
Comment (erikras) [inline docs/faq/BundleSize.md]: Oops. Missed a `react-router`. 😛

Comment (ruiaraujo) [inline docs/faq/BundleSize.md]: god dammit. Do you fix it or should I?

Comment (ruiaraujo) [inline docs/faq/BundleSize.md]: Next time I 'inspire' myself on the work of others, do cmd+R. 10 year old me would not have done this kind of mistake. 😃

Comment (ruiaraujo) [inline docs/faq/BundleSize.md]: Fixed in #2898

**Comment by erikras** [comment]
@ruiaraujo Is there some advantage to doing

```jsx
import { default as Foo } from './somedir/Foo'
```
over
```jsx
import Foo from './somedir/Foo'
```
??

The latter is less explicit, but much more readable, in my opinion.

**Comment by erikras** [comment]
Oops, nevermind. They are *EXPORT* statements. There, now we've both made absentminded mistakes on this PR. :-)

**Comment by erikras** [comment]
Ugh. ☹️ In retrospect, you really should have `git mv`'d `reduxForm.js` to `createReduxForm.js`, and then created a new `reduxForm.js`, as this PR has wiped all annotation history from the file, since all the complex code now exists in a brand new file. Oh well... live and learn...

(I probably would not have thought to do this either.  I'm just only realizing this in retrospect.)

**Comment by ruiaraujo** [comment]
Hmm, does git mv actually help with that problem?

git mv is just a mv + git add which was what I did.

**Comment by erikras** [comment]
Hmm... [So it is](http://stackoverflow.com/questions/1094269/whats-the-purpose-of-git-mv). I was under the impression that it would preserve the history of the file. I swear I've seen it do that before. Perhaps it only works if you don't immediately add another file with the old name...? Oh well.

Nevermind then. 😄

**Comment by gajus** [comment]
Is there a timeline for this getting released. :-) ?

File: docs/faq/BundleSize.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+                "redux-form": {
+                  "transform": "redux-form/es/${member}",
+                  "preventFullImport": true
+                }
+            }]
+        ]
+    }
+```
+
+##Caveat (Action Creators)
```
Comment (gajus) [inline docs/faq/BundleSize.md]: From the documentation, it is not clear whether this step is automated by `babel-plugin-transform-imports` or do I still need to use `react-router/es/actions` when using the plugin.

Comment (ruiaraujo) [inline docs/faq/BundleSize.md]: I will try to be more explicit about the problem​ then.

Comment (erikras) [inline docs/faq/BundleSize.md]: LOL, even the follow-up question didn't notice the wrong library name. 😆

**Comment by erikras** [comment]
Published as [`v6.7.0`](https://github.com/erikras/redux-form/releases/tag/v6.7.0).

**Comment by piuccio** [comment]
screen shot 2017-05-11 at 16 23 56

Interesting PR, two minor things
* You need a space in the markdown after `##`, currently it looks like in the picture above
* There's no mention on how to use this together with `immutable.js`

**Comment by ruiaraujo** [comment]
@piuccio It is the same thing to use it with immutable, just add immutable.

**Comment by piuccio** [comment]
> just add immutable

where?

`redux-form/es/whatever/immutable`
`redux-form/es/immutable/whatever`

I suppose it's the first one, but it would be nice to have it in the doc

**Comment by timhwang21** [comment]
Hi, I'm noticing that the Babel plugin doesn't work well with actions:

```
WARNING in ./src/foo/bar.js
23:23-30 "export 'default' (imported as 'actions') was not found in 'redux-form/es/actions'
```
My import:

```javascript
import { actions, Field, reduxForm } from 'redux-form';
const { reset } = actions;
```

This can be worked around from the user's end by changing the import statement, but I was wondering if it would be a good idea to add a default export to `es/actions` that contains all the actions. I think the import pattern I'm using is fairly standard, and the change would help users optimize their bundles with minimal refactoring.

Alternatively, if this is happening because I'm doing something wrong, please let me know. Thanks!

**Comment by brneto** [comment]
@erikras What is that createValues.js and values.js? Why it's not documented? Can I use it do access the form field values or you plan to remove that on the future?

**Pull request merged**